architecture/core.rst

```diff
@@ -390,6 +390,183 @@ normal world supports scheduling of processes. Even on UP systems, supporting
 efficient.
 
+Core handlers for native interrupts
+===================================
+
+OP-TEE core offers means to implement native interrupt
+(see :ref:`native_foreign_irqs`) controller drivers and means for device
+drivers to configure and register handler functions for these interrupts.
+Interrupt handlers can be nested for when an interrupt controller
+exposes interrupts which signaling is multiplexed on an interrupt
+controlled by a parent interrupt controller.
+
+Interrupt controllers are represented by an instance of ``struct itr_chip``.
+An interrupt controller exposes a given number of interrupts, each identified
+by an index from 0 to N-1 where N is the total number of interrupts exposed
+by that controller. In the literature, an interrupt index identifier
+is called interrupt number.
+
+The root interrupt controller currently implemented in OP-TEE core is a
+GIC instance for the Arm architecture.
+
+**Interrupt management API functions**
+
+Interrupt management resources are declared in header file
+kernel/interrupt.h_. Main API function are:
+
+    - ``interrupt_enable()`` and ``interrupt_disable()`` to respectively
+      enable and disable and interrupt.
+
+    - ``interrupt_mask()`` and ``interrupt_unmask()`` to respectively
+      mask/unmask and interrupt.
+
+    - ``interrupt_configure()`` to configure an interrupt detection mode
+      and the interrupt priority.
+
+    - ``interrupt_add_handler()`` and ``interrupt_remove_handler()`` to
+      respectively register and unregister an interrupt handler function
+      for a given interrupt.
+
+When configuration switch ``CFG_DRIVERS_INTERRUPT`` is enabled, core
+offers a means to use Device Tree to describe interrupts topologyin the
+platform. When used, the platform must provide a secure devicetree blob
+(DTB) which platform and device drivers can use to getreferences to
+interrupt controllers and relative interrupt numbers theyconsume.
+Main API functions are:
+
+    - ``dt_register_interrupt_provider()`` to register an interrupt
+      controller related to a given node in the Device Tree. This function
+      requires, among its arguments, a callback function that is called when
+      an interrupt consumer requires a reference to a specific interrupt
+      exposed by the controller.
+
+    - ``dt_get_interrupt()``, ``dt_get_interrupt_by_index()`` and
+      ``dt_get_interrupt_by_name()`` for an interrupt consumer to get the
+      reference to the interrupt controller and number it consumes, based
+      on the properties found in the consumer Device Tree node.
+
+**Interrupt controller drivers**
+
+Interrupt controller instance (``struct itr_chip``) registers operation
+function handlers for management of the interrupt. An interrupt controller
+must provide operation handler functions ``add``, ``enable`` and ``disable``.
+There are other operation handler functions but these are not mandatory, as
+``mask``, ``unmask``, ``rasie_pi``, ``raise_sgi`` and ``set_priority``.
```

Comment:
raise
Open question: should parentheses be used when documenting function pointers?